tests/ga/test_exthandlers.py

```diff
@@ -29,11 +34,513 @@
 from azurelinuxagent.common.utils.extensionprocessutil import TELEMETRY_MESSAGE_MAX_LEN, format_stdout_stderr, \
     read_output
 from mock import MagicMock
-from azurelinuxagent.common.protocol.wire import WireProtocol
+from azurelinuxagent.common.protocol.wire import WireProtocol, InVMArtifactsProfile
 from azurelinuxagent.ga.exthandlers import parse_ext_status, ExtHandlerInstance, get_exthandlers_handler, \
     ExtCommandEnvVariable
-from tests.tools import AgentTestCase, patch, mock_sleep, clear_singleton_instances
+from azurelinuxagent.common.protocol.extensions_config_retriever import ExtensionsConfigRetriever, \
+    GOAL_STATE_SOURCE_FABRIC, GOAL_STATE_SOURCE_FASTTRACK, _GOAL_STATE_SOURCE_FILE_NAME, _SEQUENCE_NUMBER_FILE_NAME, \
+    _INCARNATION_FILE_NAME
+from tests.protocol.mocks import MockWireClient, MockProtocol
+from tests.protocol.mockwiredata import WireProtocolData, DATA_FILE, DATA_FILE_FAST_TRACK_NO_SETTINGS
+from tests.tools import AgentTestCase, patch, mock_sleep, clear_singleton_instances, i_am_root
+
+
+class TestGoalState(AgentTestCase):
+
+    def setUp(self):
+        super(TestGoalState, self).setUp()
+        # Since ProtocolUtil is a singleton per thread, we need to clear it to ensure that the test cases do not
+        # reuse a previous state
+        clear_singleton_instances(ProtocolUtil)
 
+    def test_set_fabric(self):
```

Comment:
As part of our testing fundamentals, we're trying to avoid using private functions for our tests and only mock the public interfaces. This is to ensure that we test the uber functionality of the component rather than specific pieces of it. Additionally, adding tests for internal components leads to humongous pains if you plan on changing some internals later (as we've seen so frequently with the Linux Guest Agent code).

For example: instead of testing `test_set_fabric` and `test_set_fasttrack`, you could maybe add some test 
 like `test_extension_config_retriever_commit_processed_successfully` or something without mocking any internals. 

With this in mind, could you please refactor some of your tests to mainly test the functionalities (what its supposed to do) rather than testing the internal components (how its doing it). 

We can hop on a call if you have any questions about my ask here. 
Thanks a lot for doing this!

Reply:
discussed on the call. I did remove a few tests, but IMHO the rest are important enough to leave, even though the logic should be covered elsewhere.